src/eos-updater-apply.c

```diff
@@ -73,41 +77,592 @@ get_test_osname (void)
+                            &local_error))
+    {
+      if (!g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return NULL;
+        }
+
+      /* If the file wasn't found, assume that no applications have been
+       * explicitly uninstalled */
+      return g_ptr_array_new ();
+    }
+
+  transactions = g_ptr_array_new_full (0, (GDestroyNotify) json_object_unref);
+  parser = json_parser_new ();
+
+  contents_sep = contents;
+  while ((line = strsep (&contents_sep, "\n")) != NULL)
+    {
+      JsonNode *node = NULL;
+
+      if (!json_parser_load_from_data (parser, line, -1, error))
+        return FALSE;
+
+      node = json_parser_get_root (parser);
+
+      /* If there's no node, it probably meant this line was empty, so
+       * just ignore it */
+      if (!node)
+        continue;
+
+      g_ptr_array_add (transactions, json_node_dup_object (node));
+    }
+
+  return g_steal_pointer (&transactions);
+}
+
+static GHashTable *
+determine_set_of_explictly_uninstalled_applications (FlatpakInstallation *installation,
+                                                     GCancellable        *cancellable,
+                                                     GError              **error)
+{
+  g_autoptr(GFile) installation_dir = flatpak_installation_get_path (installation);
+  g_autoptr(GFile) transaction_log_file = g_file_get_child (installation_dir,
+                                                            "transactions.log");
+  g_autoptr(GPtrArray) transactions = transactions_from_log_file (transaction_log_file,
+                                                                  error);
+
+  if (!transactions)
+    return NULL;
+
+  return determine_set_of_explicitly_uninstalled_applications_from_flatpak_log (transactions, error);
+}
+
+static GPtrArray *
+resolve_flatpaks_to_deploy (GPtrArray    *flatpaks_this_commit_wants,
+                            GCancellable *cancellable,
+                            GError       **error)
+{
+  g_autoptr(FlatpakInstallation) installation = get_flatpak_installation (cancellable, error);
+  g_autoptr(GPtrArray) resolved_flatpaks = g_ptr_array_new_with_free_func (g_object_unref);
+  g_autoptr(GHashTable) explicitly_removed_flatpaks = NULL;
+  g_autoptr(GHashTable) already_installed_flatpaks = NULL;
+  gsize i = 0;
+
+  explicitly_removed_flatpaks = determine_set_of_explictly_uninstalled_applications (installation, cancellable, error);
+
+  if (!explicitly_removed_flatpaks)
+    return NULL;
+
+  already_installed_flatpaks = determine_set_of_installed_applications (installation, cancellable, error);
+
+  if (!already_installed_flatpaks)
+    return NULL;
+
+  /* Filter the flatpaks this commit wants by removing any flatpaks
+   * which were already explicitly removed earlier, or which are already
+   * installed */
+  for (; i < flatpaks_this_commit_wants->len; ++i)
```

Comment:
Please move the initialisation of `i` into the `for` loop to make the loop bounds more obvious.